modules/indexing/src/main/java/org/apache/ignite/internal/visor/verify/ValidateIndexesClosure.java

```diff
@@ -788,4 +836,269 @@ else if (e.getCause() instanceof IgniteException)
+                    return new T2<>(null, 0L);
+
+                String cacheName = cacheCtx.name();
+                String tblName = idx.getTable().getName();
+                String idxName = idx.getName();
+
+                try {
+                    long indexSize = ignite.context().query().getIndexing().indexSize(cacheName, tblName, idxName);
+                    return new T2<>(null, indexSize);
+                }
+                catch (Throwable t) {
+                    Throwable idxSizeErr = new IgniteException("Index size calculation error [" +
+                        cacheGrpInfo(cacheCtx.group()) + ", " + cacheInfo(cacheCtx) + ", tableName=" +
+                        tblName + ", idxName=" + idxName + ", err=" + t.getMessage() + "]", t);
+
+                    error(log, idxSizeErr);
+
+                    return new T2<>(idxSizeErr, 0L);
+                }
+            }
+            finally {
+                processedIdxSizes.incrementAndGet();
+
+                printProgressOfIndexValidationIfNeeded();
+            }
+        });
+    }
+
+    /**
+     * Return cache group info string.
+     *
+     * @param cacheGrpCtx Cache group context.
+     * @return Cache group info string.
+     */
+    private String cacheGrpInfo(CacheGroupContext cacheGrpCtx) {
+        return "cacheGrpName=" + cacheGrpCtx.name() + ", cacheGrpId=" + cacheGrpCtx.groupId();
+    }
+
+    /**
+     * Return cache info string.
+     *
+     * @param cacheCtx Cache context.
+     * @return Cache info string.
+     */
+    private String cacheInfo(GridCacheContext cacheCtx) {
+        return "cacheName=" + cacheCtx.name() + ", cacheId=" + cacheCtx.cacheId();
+    }
+
+    private void checkSizes(
+        List<T3<CacheGroupContext, GridDhtLocalPartition, Future<CacheSize>>> cacheSizesFutures,
+        List<T3<GridCacheContext, Index, Future<T2<Throwable, Long>>>> idxSizeFutures,
+        Map<String, ValidateIndexesCheckSizeResult> checkSizeResult
+    ) throws ExecutionException, InterruptedException {
+        if (!checkSizes)
+            return;
+
+        Map<Integer, CacheSize> cacheSizeTotal = new HashMap<>();
+
+        for (T3<CacheGroupContext, GridDhtLocalPartition, Future<CacheSize>> cacheSizeFut : cacheSizesFutures) {
+            CacheGroupContext cacheGrpCtx = cacheSizeFut.get1();
+            CacheSize cacheSize = cacheSizeFut.get3().get();
+
+            Throwable cacheSizeErr = cacheSize.err;
+
+            int grpId = cacheGrpCtx.groupId();
+
+            if (failCalcCacheSizeGrpIds.contains(grpId) && nonNull(cacheSizeErr)) {
+                checkSizeResult.computeIfAbsent(
+                    cacheGrpInfo(cacheGrpCtx),
+                    s -> new ValidateIndexesCheckSizeResult(0, new ArrayList<>())
+                ).issues().add(new ValidateIndexesCheckSizeIssue(null, 0, cacheSizeErr));
+            } else {
+                cacheSizeTotal.computeIfAbsent(grpId, i -> new CacheSize(null, new HashMap<>()))
```

Comment:
I assume you can to map index name to size, instead that complex matching.

Reply:
The index name is not unique, so I can 't, for example PK index.